### PR TITLE
KTOR-829 Support `100 Continue` on client side

### PR DESCRIPTION
**Pull request by marychatte**
Client
[KTOR-829](https://youtrack.jetbrains.com/issue/KTOR-829/Support-100-Continue)
[rfc7231 section-5.1.1](https://www.rfc-editor.org/rfc/rfc7231#section-5.1.1)

### CODE REVIEW

File: ktor-client/ktor-client-cio/jvmAndNix/src/io/ktor/client/engine/cio/Endpoint.kt [hunk 1]
```diff
@@ -119,13 +121,68 @@ internal class Endpoint(
...
+    private suspend fun processExpectContinue(
+        request: HttpRequestData,
+        input: ByteReadChannel,
+        output: ByteWriteChannel,
+        originOutput: ByteWriteChannel,
+        callContext: CoroutineContext,
+        requestTime: GMTDate,
+        overProxy: Boolean,
+    ) = withContext(callContext) {
+        writeHeaders(request, output, overProxy, closeChannel = false)
```
Comment (rsinukov) [inline ktor-client/ktor-client-cio/jvmAndNix/src/io/ktor/client/engine/cio/Endpoint.kt:147]: why setting `closeChannel` to false?

File: ktor-client/ktor-client-cio/jvmAndNix/src/io/ktor/client/engine/cio/Endpoint.kt [hunk 2]
```diff
@@ -119,13 +121,68 @@ internal class Endpoint(
...
+                        headers.remove(HttpHeaders.Expect)
+                    }.build()
+                    writeRequest(newRequest, output, callContext, overProxy)
+                }
+
+                HttpStatusCode.Continue -> {
+                    writeBody(request, output, callContext)
+                }
+
+                else -> return@withContext response
```
Comment (rsinukov) [inline ktor-client/ktor-client-cio/jvmAndNix/src/io/ktor/client/engine/cio/Endpoint.kt:177]: `body` should be canceled here to avoid leaks

File: ktor-client/ktor-client-cio/jvmAndNix/src/io/ktor/client/engine/cio/Endpoint.kt [hunk 3]
```diff
@@ -159,7 +216,7 @@ internal class Endpoint(
                 }
 
                 val socket = when (connectTimeout) {
-                    HttpTimeout.INFINITE_TIMEOUT_MS -> connect()
+                    HttpTimeout.INFINITE_TIMEOUT_MS -> coroutineScope { connect() }
```
Comment (rsinukov) [inline ktor-client/ktor-client-cio/jvmAndNix/src/io/ktor/client/engine/cio/Endpoint.kt:219]: why this change?

File: ktor-client/ktor-client-cio/jvmAndNix/src/io/ktor/client/engine/cio/utils.kt [hunk 4]
```diff
@@ -148,10 +176,11 @@ internal suspend fun readResponse(
 
         val body = when {
             request.method == HttpMethod.Head ||
-                status in listOf(HttpStatusCode.NotModified, HttpStatusCode.NoContent) ||
+                status in listOf(HttpStatusCode.NotModified, HttpStatusCode.NoContent, HttpStatusCode.Continue) ||
                 status.isInformational() -> {
```
Comment (rsinukov) [inline ktor-client/ktor-client-cio/jvmAndNix/src/io/ktor/client/engine/cio/utils.kt:180]: `isInformational` covers `HttpStatusCode.Continue`

**Review by rsinukov — CHANGES_REQUESTED** [review]
some comments

File: ktor-client/ktor-client-cio/jvmAndNix/src/io/ktor/client/engine/cio/Endpoint.kt (hunk 3 above)
Comment (marychatte) [reply ktor-client/ktor-client-cio/jvmAndNix/src/io/ktor/client/engine/cio/Endpoint.kt:219]: It was a warning and [compilation error](https://ktor.teamcity.com/buildConfiguration/Ktor_JPMSCheck/187189?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildProblemsSection=true&expandBuildChangesSection=true) because of it

File: ktor-client/ktor-client-cio/jvmAndNix/src/io/ktor/client/engine/cio/Endpoint.kt (hunk 1 above)
Comment (marychatte) [reply ktor-client/ktor-client-cio/jvmAndNix/src/io/ktor/client/engine/cio/Endpoint.kt:147]: Should be true, thanks

File: ktor-client/ktor-client-cio/jvmAndNix/src/io/ktor/client/engine/cio/Endpoint.kt [hunk 5]
```diff
@@ -119,13 +121,71 @@ internal class Endpoint(
...
+            coroutineContext.job.invokeOnCompletion(onCancelling = true) {
+                if (it == null) job.complete()
+                else job.completeExceptionally(it)
+            }
+            callContext.job.invokeOnCompletion {
+                if (it == null) job.complete()
+                else job.completeExceptionally(it)
+            }
+
+            readResponse(requestTime, request, input, originOutput, callContext + job)
```
Comment (rsinukov) [inline ktor-client/ktor-client-cio/jvmAndNix/src/io/ktor/client/engine/cio/Endpoint.kt:160]: It's better to have `awaitContent` instead of `readResponse` inside `withTimeoutOrNull`. otherwise, you can consume only part of the response until the timeout is reached and the channel will have partial data remaining.

**Review by rsinukov — CHANGES_REQUESTED** [review]
@e5l Please also take a look at it

File: ktor-client/ktor-client-cio/jvmAndNix/src/io/ktor/client/engine/cio/Endpoint.kt (hunk 3 above)
Comment (e5l) [reply ktor-client/ktor-client-cio/jvmAndNix/src/io/ktor/client/engine/cio/Endpoint.kt:213]: It looks like a bug in the report, could you log the issue in Kotlin YT and add `@Suppress("UNUSED_EXPRESSION")` so we can remove `coroutineScope` here

**Review by e5l — APPROVED** [review]

File: ktor-client/ktor-client-cio/jvmAndNix/src/io/ktor/client/engine/cio/Endpoint.kt (hunk 3 above)
Comment (marychatte) [reply ktor-client/ktor-client-cio/jvmAndNix/src/io/ktor/client/engine/cio/Endpoint.kt:213]: Created issue [KT-57967](https://youtrack.jetbrains.com/issue/KT-57967)

**Review by rsinukov — APPROVED** [review]
